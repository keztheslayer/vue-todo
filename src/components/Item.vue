<template>
    <div
        class="todo__list-item"
        :class="isChecked ? 'todo__list-item_state_finished' : ''"
    >
        <div class="todo__list-item-wrapper">
            <label class="todo__label">
                <input
                    v-model="isChecked"
                    class="todo__checkbox"
                    type="checkbox"
                    @click="checkItem( index )"
                >
                <span class="todo__list-item-text">{{ text }}</span>
            </label>
            <button
                class="todo__list-item-edit"
                @click="editTask( index )"
            />
            <button
                type="button"
                class="close todo__remove-btn"
                @click="removeTask( index )"
            >
                <span>&times;</span>
            </button>
        </div>
        <hr v-if="hideDivider">
    </div>
</template>

<script>
import { mapActions } from 'vuex';

export default {
    props : {
        isChecked : {
            type    : Boolean,
            default : false,
        },
        text : {
            type    : String,
            default : '',
        },
        index : {
            type    : Number,
            default : 0,
        },
        hideDivider : {
            type    : Boolean,
            default : true,
        },
    },
    methods : {
        ...mapActions( [
            'editTask',
            'checkItem',
            'removeTask',
        ] ),
    },
};
</script>

<style lang="less">
.todo {
    &__label {
        cursor: pointer;
        color: #333;
        font-size: 18px;
        margin: 0;
        display: block;
        flex-grow: 1;
    }

    &__list-item {
        position: relative;
        margin-bottom: 6px;

        &:last-child {
            margin-bottom: 0;
        }
    }

    &__list-item-text {
        transition: .15s ease-in;
    }

    &__list-item-wrapper {
        display: flex;
        flex-wrap: nowrap;
        justify-content: space-between;
        align-items: center;
    }

    &__list-item-edit {
        display: block;
        width: 24px;
        height: 24px;
        background: transparent;
        border: none;
        appearance: none;
        margin-right: 12px;

        &:focus {
            outline: none;
        }
    }

    &__list-item-edit:before {
        content: '\f040';
        cursor: pointer;
        color: #000;
        font: 15px "fontAwesome";
        line-height: 1;
        width: 1em;
        display: inline-block;
        opacity: 0.5;
        transition: opacity 0.3s;
    }
    &__list-item-edit:hover:before {
        opacity: 1;
    }

    &__checkbox + &__list-item-text:hover:before {
        color: #fe4365;
    }

    &__checkbox {
        display: none;
    }

    &__checkbox + &__list-item-text:before {
        content: "";
        color: #dddfe6;
        font-family: "fontAwesome";
        line-height: 1;
        width: 1em;
        display: inline-block;
        margin-right: 8px;
    }

    &__checkbox:checked + &__list-item-text:before {
        content: "";
        color: #fe4365;
        animation: tick 150ms ease-in;
    }

    &__checkbox:checked + &__list-item-text {
        color: #7e7e7e;
    }

    &__checkbox:disabled + &__list-item-text:before {
        content: "";
        color: #dddfe6;
    }

    &__remove-btn {
        display: block;
        width: 24px;
        height: 24px;

        &:focus {
            outline: none;
        }
    }
}

@keyframes tick {
    0% {
        transform: scale(0);
    }
    90% {
        transform: scale(1.3);
    }
    100% {
        transform: scale(1);
    }
}
</style>

