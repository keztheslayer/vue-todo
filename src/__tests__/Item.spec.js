import Item from '../components/Item.vue';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';

const localVue = createLocalVue();

const checkedTaskWithDivider = {
    isChecked   : true,
    text        : 'Checked task',
    hideDivider : false,
};

const uncheckedTaskWithoutDivider = {
    isChecked   : false,
    text        : 'Unchecked task',
    hideDivider : true,
};

localVue.use( Vuex );

describe( 'Item.vue', () => {
    it( 'is a Vue instance', () => {
        const wrapper = mount( Item );

        wrapper.vm.$forceUpdate();

        expect( wrapper.isVueInstance() ).toBeTruthy();
    } );

    it( 'renders unchecked task with divider', () => {
        const wrapper = mount( Item, {
            propsData : checkedTaskWithDivider,
        } );

        expect( wrapper.element ).toMatchSnapshot();
    } );

    it( 'renders unchecked task without divider', () => {
        const wrapper = mount( Item, {
            propsData : uncheckedTaskWithoutDivider,
        } );

        expect( wrapper.element ).toMatchSnapshot();
    } );
} );

describe( 'Item\'s Vuex actions', () => {
    let actions;

    let store;

    beforeEach( () => {
        actions = {
            removeTask : jest.fn(),
            editTask   : jest.fn(),
        };
        store = new Vuex.Store( {
            actions,
        } );
    } );

    it( 'dispatches "removeTask" when clicking on remove-btn', () => {
        const wrapper = mount(
            Item,
            {
                store,
                localVue,
            }, );
        const removeButton = wrapper.find('.todo__remove-button');

        removeButton.trigger('click');
        expect( actions.removeTask ).toHaveBeenCalled();
    } );

    it( 'dispatches "editTask" when clicking on edit-button', () => {
        const wrapper = mount(
            Item,
            {
                store,
                localVue,
            }, );
        const editButton = wrapper.find('.todo__edit-button');

        editButton.trigger('click');
        expect( actions.editTask ).toHaveBeenCalled();
    } );
} );
