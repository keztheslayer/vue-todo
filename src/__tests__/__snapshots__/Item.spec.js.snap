// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item.vue renders unchecked task with divider 1`] = `
<div
  class="todo__list-item todo__list-item_state_finished"
>
  <div
    class="todo__list-item-wrapper"
  >
    <label
      class="todo__label"
    >
      <input
        class="todo__checkbox"
        type="checkbox"
      />
       
      <span
        class="todo__list-item-text"
      >
        Checked task
      </span>
    </label>
     
    <button
      class="todo__edit-button"
    />
     
    <button
      class="close todo__remove-button"
      type="button"
    >
      <span>
        ×
      </span>
    </button>
  </div>
   
  <!---->
</div>
`;

exports[`Item.vue renders unchecked task without divider 1`] = `
<div
  class="todo__list-item"
>
  <div
    class="todo__list-item-wrapper"
  >
    <label
      class="todo__label"
    >
      <input
        class="todo__checkbox"
        type="checkbox"
      />
       
      <span
        class="todo__list-item-text"
      >
        Unchecked task
      </span>
    </label>
     
    <button
      class="todo__edit-button"
    />
     
    <button
      class="close todo__remove-button"
      type="button"
    >
      <span>
        ×
      </span>
    </button>
  </div>
   
  <hr />
</div>
`;
